packages/zoltan2/src/algorithms/partition/Zoltan2_AlgMultiJagged.hpp

```diff
@@ -54,7037 +54,8852 @@
-                        break;
-                    arr[i+1] = arr[i];
-                }
-                arr[i+1]=a;
-            }
-            if (jstack == 0)
-                break;
-            ir=istack[jstack--];
-            l=istack[jstack--];
-        }
-        else
-        {
-            k=(l+ir) >> 1;
-
-            ZOLTAN2_ALGMULTIJAGGED_SWAP(arr[k],arr[l+1], temp)
-            if (arr[l+1].val > arr[ir].val)
-            {
-                ZOLTAN2_ALGMULTIJAGGED_SWAP(arr[l+1],arr[ir],temp)
-            }
-            if (arr[l].val > arr[ir].val)
-            {
-                ZOLTAN2_ALGMULTIJAGGED_SWAP(arr[l],arr[ir],temp)
-            }
-            if (arr[l+1].val > arr[l].val)
-            {
-                ZOLTAN2_ALGMULTIJAGGED_SWAP(arr[l+1],arr[l],temp)
-            }
-            i=l+1;
-            j=ir;
-            a=arr[l];
-            aval = a.val;
-            for (;;)
-            {
-                do i++; while (arr[i].val < aval);
-                do j--; while (arr[j].val > aval);
-                if (j < i) break;
-                ZOLTAN2_ALGMULTIJAGGED_SWAP(arr[i],arr[j],temp);
-            }
-            arr[l]=arr[j];
-            arr[j]=a;
-            jstack += 2;
-            if (jstack > NSTACK){
-                std::cout << "uqsort: NSTACK too small in sort." << std::endl;
-                exit(1);
-            }
-            if (ir-i+1 >= j-l)
-            {
-                istack[jstack]=ir;
-                istack[jstack-1]=i;
-                ir=j-1;
-            }
-            else
-            {
-                istack[jstack]=j-1;
-                istack[jstack-1]=l;
-                l=i;
-            }
+void uqsort(IT n, uSortItem<IT, WT> * arr) {
+  int NSTACK = 50;
+  int M = 7;
+  IT         i, ir=n, j, k, l=1;
+  IT         jstack=0, istack[50];
+  WT aval;
+  uSortItem<IT,WT>    a;
+
+  --arr;
+  for(;;) {
+    if(ir-l < M) {
+      for(j=l+1;j<=ir;j++) {
+        a=arr[j];
+        aval = a.val;
+        for(i=j-1;i>=1;i--) {
+          if(arr[i].val <= aval)
+            break;
+          arr[i+1] = arr[i];
         }
+        arr[i+1]=a;
+      }
+      if(jstack == 0)
+          break;
+      ir=istack[jstack--];
+      l=istack[jstack--];
+    }
+    else {
+      k=(l+ir) >> 1;
+      std::swap(arr[k],arr[l+1]);
+      if(arr[l+1].val > arr[ir].val) {
+        std::swap(arr[l+1],arr[ir]);
+      }
+      if(arr[l].val > arr[ir].val) {
+        std::swap(arr[l],arr[ir]);
+      }
+      if(arr[l+1].val > arr[l].val) {
+        std::swap(arr[l+1],arr[l]);
+      }
+      i=l+1;
+      j=ir;
+      a=arr[l];
+      aval = a.val;
+      for(;;) {
+        do i++; while (arr[i].val < aval);
+        do j--; while (arr[j].val > aval);
+        if(j < i) break;
+        std::swap(arr[i],arr[j]);
+      }
+      arr[l]=arr[j];
+      arr[j]=a;
+      jstack += 2;
+      if(jstack > NSTACK) {
+        std::cout << "uqsort: NSTACK too small in sort." << std::endl;
+        exit(1);
```

Comment:
It would be better at least to call `MPI_Abort` or `std::terminate` so that we have some opportunity to print a stack trace etc.

Reply:
switched these exit(1) calls to std::terminate

Reply:
or you could just call `std::sort` :-)